### PR TITLE
Refactor sky regions

### PR DESCRIPTION
**Pull request by astrofrog**
This changes the current sky regions to be simply regions that represent the same shapes as the pixel regions but with properties specified in world coordinates (e.g. a circle in an image with a center given by a skycoord) and for now this removes the sphericalskyregion functionality for circles (will add this back after). This is to address https://github.com/astropy/regions/issues/107

This also implements some of the missing area calculations, to_shapely, to_sky, etc and I've split up the docs into several sections and improved it in places.

You can preview the updated docs at: http://astropy-regions-astrofrog.readthedocs.io/en/latest/

I plan to then try and add the full spherical sky polygon and circle in a separate PR, but as-is this should now provide full compatibility/consistency with ds9.

### CODE REVIEW

**Comment by astrofrog** [comment]
@joleroi - here's the file I was using to run the integration test: https://www.dropbox.com/sh/pfh42u7mexazuo8/AABn40AUsgPB785g58FHZhiQa?dl=0

**Comment by astrofrog** [comment]
Just FYI for anyone looking at this, we still need to do some work to check that the ellipse and rectangle are correctly rotated (they appear to be 90 degrees off at the moment compared to ds9)

File: regions/core/core.py [hunk 1]
```diff
@@ -283,47 +259,29 @@ def union(self, other):
         from .compound import CompoundSkyRegion
         return CompoundSkyRegion(self, other, operator.or_)
 
-    @abc.abstractmethod
-    def contains(self, skycoord):
+    def contains(self, skycoord, wcs):
```
Comment (joleroi) [inline regions/core/core.py]: Can we add ``__contains__`` calling this method also for SkyRegions (it's there for pixel regions)

Comment (joleroi) [inline regions/core/core.py]: Ah no you can't because you have to pass a ``wcs`` ...

File: regions/core/pixcoord.py [hunk 2]
```diff
@@ -143,6 +143,7 @@ def from_sky(cls, skycoord, wcs, origin=_DEFAULT_WCS_ORIGIN, mode=_DEFAULT_WCS_M
         Calls :meth:`astropy.coordinates.SkyCoord.to_pixel`.
         See parameter description there.
         """
+        print(skycoord)
```
Comment (joleroi) [inline regions/core/pixcoord.py]: remove?

File: regions/core/core.py (hunk 1 above)
Comment (astrofrog) [inline regions/core/core.py]: Yep!

File: regions/core/pixcoord.py (hunk 2 above)
Comment (astrofrog) [inline regions/core/pixcoord.py]: Oops, thanks!

**Comment by astrofrog** [comment]
I think this is basically ready for review at this point. I need to find a better way to do the Point patch, but I think that can be fixed later. There are also some FIXMEs and TODOs but none are critical but more for future.

File: regions/core/core.py [hunk 3]
```diff
@@ -283,47 +259,29 @@ def union(self, other):
...
+        Check whether a sky coordinate falls inside the region
 
         Parameters
         ----------
         skycoord : `~astropy.coordinates.SkyCoord`
             The position or positions to check
+        wcs : `~astropy.wcs.WCS` instance
+            The world coordinate system transformation to assume
         """
+        from .pixcoord import PixCoord
```
Comment (cdeil) [inline regions/core/core.py]: Move import to the top of the file.

**Comment by cdeil** [comment]
The docs restructuring looks good to me. The section that sticks out now even more as useless is "What next?". I guess it could be removed. Or rewritten, e.g. the obvious thing to do next is look at the API docs for details, so that could be mentioned.

File: regions/core/core.py (hunk 3 above)
Comment (astrofrog) [inline regions/core/core.py]: I can't, because of circular imports

**Comment by astrofrog** [comment]
Note that this now implements the change of ellipse parameters to width/height as discussed in https://github.com/astropy/regions/issues/86

**Comment by astrofrog** [comment]
@cdeil - I'll work on making sure we have proper test for all the rotations

File: regions/core/core.py (hunk 3 above)
Comment (cdeil) [inline regions/core/core.py]: Ah, OK, I see we have `from .core import _DEFAULT_WCS_ORIGIN, _DEFAULT_WCS_MODE` in `pixcoord.py`.  This is not nice that we have circular imports between `pixcoord.py` and `core.py`. I'm +1 to move those constants to `pixcoord.py` to have it be standalone (and because all pixel / world transforms should go through `PixCoord` to abstract away this choice we've made for `regions`.

**Comment by astrofrog** [comment]
One of the easiest ways to make sure that everything is consistent is to make plots like this:

everything

I was wondering whether we might want to consider adding image tests to the test suite to test this kind of all-ecompasing plot? Or would people prefer that we really just compare numerical results? (no strong feelings, but just want to know before implementing anything)

**Comment by keflavich** [comment]
I like this sort of visual test, but I would prefer that it be kept optional so that I can run the tests locally without getting popup windows and faster.

**Comment by astrofrog** [comment]
Ok - the main difficulty with testing this without plots is how to test e.g. that the shapely region is correct. I don't really want to have to write out all the array values for the ellipse edges. So what about adding a few image tests (keeping it to a minimum of one plot per shape), exclude axis labels (since that's usually what changes between matplotlib versions), making it so when tests are run the backend is set to Agg (can be done) to avoid pop-up windows? The tests should still be speedy.

**Comment by astrofrog** [comment]
Another possibility is to simply write out the exact ellipse vertex arrays to ASCII files (like we do for masks) but that produces less visual checks.

**Comment by cdeil** [comment]
+1 to separate test for computations (e.g. contains) from plotting tests.

@astrofrog - you mention that it's difficult to write the test for the Shapely objects.
Why do we need large arrays / masks there?
Can't you assert on the data member parameters of the Shapely ellipse object, and maybe in addition call contains on ~ 3 well-chose points that are inside and outside and establish the ellipse orientation?

**Comment by astrofrog** [comment]
I was being lazy... I will implement proper tests, as I have some ideas on doing that efficiently

File: regions/core/core.py (hunk 3 above)
Comment (astrofrog) [inline regions/core/core.py]: Let's do that in a separate PR? This one is getting out of control :)

**Comment by astrofrog** [comment]
Ok, so if tests pass, I would suggest we merge this and we can always tweak things in future PRs. This one is getting a bit out of control. I've refactored the tests to move a lot of the common logic to common base classes, and I've made it so that we test our contains method against shapely for hundreds of points (which gives me more confidence). We could still add more tests but I would prefer to merge first and add these in separate PRs.

File: docs/ds9.rst [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    >>> write_ds9(regions, filename)
+    >>> regions = read_ds9(filename)
+    >>> regions
+    [CircleSkyRegion
+     center: <FK5 Coordinate (equinox=J2000.000): (ra, dec) in deg
+         (245.3477, 24.4291)>
+     radius: 3.0 deg]
+
+Often DS9 region files contain extra information about colors or other attributes.
+This information is lost when converting to `~regions.Region` objects.
```
Comment (keflavich) [inline docs/ds9.rst]: This should not be true; the data are (or at least were) preserved in `.meta` and `.visual`

Comment (astrofrog) [inline docs/ds9.rst]: I don't want to modify this as it might conflict with changes in #130, so I think we should change it there once this PR is merged.

**Comment by astrofrog** [comment]
@cdeil - any objections to merging this now? I plan to work on further improvements, but this PR is getting a little out of hand and already implements the original refactoring plan  (we should get #130 merged soon too)

**Comment by cdeil** [comment]
Sorry, I didn't know what I was supposed to merge. Merging now.
(we can do further review in master and everyone that has time can do follow-up issues or PRs)

**Comment by cdeil** [comment]
@astrofrog - Thank you!!!!

File: docs/ds9.rst (hunk 4 above)
Comment (joleroi) [inline docs/ds9.rst]: Fixed in #130

**Comment by keflavich** [comment]
@astrofrog This is a too-late comment, but I have a problem with the new API.  I have a set of regions that have been created as sky regions with no associated WCS: they are positions on the sky with circles around them representing regions of constant angular distance.  When I do `reg.contains(coordinate)`, it make no sense at all to require a WCS.

While I agree with the original premise, since many regions are undefinable without an associated projection (e.g., rectangles), I expect other users to encounter this case commonly as well.  The workaround now is hideous: I have to create an arbitrary WCS from scratch.  Do you have any alternative suggestions, or ideas how we can make this more sensible?

**Comment by cdeil** [comment]
We discussed this quite a bit, and my understanding is this:
* There's very few sky regions where the shape (e.g. "contains") is well-defined without a WCS. I'm only aware of circle and polygon.
* The plan is to bring that "true sky region" functionality back as separate classes.
* Of course trying to add the "true sky region" case back into the existing circle and polygon sky region classes can be discussed again, but I think it's messy. In any case, there's some things that aren't so nice, e.g. the "true sky regions" aren't available in the DS9 serialisation format, i.e. DS9 just doesn't support them.

**Comment by keflavich** [comment]
Yeah, it might only make sense to support those two regions, but we should.

In the circle case, the true sky region and WCS-based region approach should be equivalent, right?  Is it possible that the true sky region approach is also more computationally efficient?

**Comment by astrofrog** [comment]
I'm going to add back the 'spherical' circle and polygon in a future PR.

> In the circle case, the true sky region and WCS-based region approach should be equivalent, right? Is it possible that the true sky region approach is also more computationally efficient?

No, a true spherical circle is not necessarily equivalent to the current approach, especially for images that span a large area of sky. But checking for points inside the true spherical circle might indeed be more efficient, though that remains to be tested.

**Comment by joleroi** [comment]
> In the circle case, the true sky region and WCS-based region approach should be equivalent, right?

For completeness: I guess this was discussed  in https://github.com/astropy/regions/issues/76

**Comment by bsipocz** [comment]
This PR splitting up the docs has a side effect of now the examples are failing, and the references in the narrative doesn't make sense (https://astropy-regions.readthedocs.io/en/latest/contains.html)

**Comment by cdeil** [comment]
ref #263
